sdks/java/core/src/main/java/org/apache/beam/sdk/coders/RowCoderGenerator.java

```diff
@@ -336,27 +387,30 @@ public InstrumentedType prepare(InstrumentedType instrumentedType) {
 
     // The decode method of the generated Coder delegates to this method to evaluate all of the
     // per-field Coders.
-    static Row decodeDelegate(Schema schema, Coder[] coders, InputStream inputStream)
+    static Row decodeDelegate(
+        Schema schema, Coder[] coders, int[] encodingPosToIndex, InputStream inputStream)
         throws IOException {
       int fieldCount = VAR_INT_CODER.decode(inputStream);
 
       BitSet nullFields = NULL_LIST_CODER.decode(inputStream);
-      List<Object> fieldValues = Lists.newArrayListWithCapacity(coders.length);
-      for (int i = 0; i < fieldCount; ++i) {
+      Object[] fieldValues = new Object[coders.length];
+      for (int encodingPos = 0; encodingPos < fieldCount; ++encodingPos) {
+        int rowIndex = encodingPosToIndex[encodingPos];
         // In the case of a schema change going backwards, fieldCount might be > coders.length,
         // in which case we drop the extra fields.
```

Comment:
Is there anything in `RowCoderTest` exercising the `i >= coders.length` case? I've always been a little confused about what would trigger it. I wouldn't think it would be possible for the old coder to read data from the new coder. Maybe we should exercise removing a field from the end of the field list?

Reply:
Added test. This is mostly so that if people emergency rollback updates things don't start crashing on them. (ideally you would roll back to a previous state snapshot, but in emergency situations it's nice that something works)